### PR TITLE
Temporary memory leak in `FutureImpl.executors`

### PR DESCRIPTION
**Pull request by jglick**
Alarmed by [this thread](https://community.jenkins.io/t/jenkins-memory-leak/10309/5?u=jglick) I tried to see if I could reproduce a memory leak from running K8s builds with WebSocket. I could not reproduce any major leak, but did find some minor issues.

### Testing done

Without patch, the test fails with this reference chain:

```
private static jenkins.model.Jenkins jenkins.model.Jenkins.theInstance->
hudson.model.Hudson@39f8cf44-queue->
hudson.model.Queue@d78af81-leftItems->
com.google.common.cache.LocalCache$LocalManualCache@7a55a792-localCache->
com.google.common.cache.LocalCache@4816e261-segments->
[Lcom.google.common.cache.LocalCache$Segment;@de1aeaf-[1]->
com.google.common.cache.LocalCache$Segment@619f68f4-table->
java.util.concurrent.atomic.AtomicReferenceArray@5d30ba49-array->
[Ljava.lang.Object;@251e4e2e-[2]->
com.google.common.cache.LocalCache$StrongWriteEntry@e03c78d-valueReference->
com.google.common.cache.LocalCache$StrongValueReference@586c2859-referent->
hudson.model.Queue$LeftItem@6425b839-future->
hudson.model.queue.FutureImpl@7665f1fd-executors->
java.util.HashSet@76f2bac1-map->
java.util.HashMap@1ebad322-table->
[Ljava.util.HashMap$Node;@49eb0aac-[2]->
java.util.HashMap$Node@5f7e9273-key->
hudson.model.Executor@94d52cf-owner->
hudson.slaves.SlaveComputer@65397da8
```

`LeftItem`s are kept around for 5m.

### Proposed changelog entries

- Some agent-related objects could be kept in memory after being disconnected and removed from the computer list.

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: test/pom.xml
Comment (jglick) [inline test/pom.xml:124]: https://github.com/jenkinsci/jenkins-test-harness/pull/669

File: test/src/test/java/hudson/model/ComputerTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/ComputerTest.java:277]: Just in case there was some problem with the `Channel` separately from the `Computer`. I could not reproduce any in this test, other than https://github.com/jenkinsci/remoting/pull/694 which was reproduced interactively.

**Review by NotMyFault — APPROVED** [review]

File: test/src/test/java/hudson/model/ComputerTest.java
Comment (timja) [inline test/src/test/java/hudson/model/ComputerTest.java:270]: may not pass on windows `ComputerTest` failed in https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-8640/3/pipeline/235

Comment (jglick) [reply test/src/test/java/hudson/model/ComputerTest.java:270]: Hmm, though `junit` did not catch that for some reason.

File: core/src/main/java/hudson/model/queue/FutureImpl.java
Comment (basil) [inline core/src/main/java/hudson/model/queue/FutureImpl.java:100]: https://github.com/jenkinsci/jenkins-test-harness/pull/643#discussion_r1304946075

Comment (jglick) [reply core/src/main/java/hudson/model/queue/FutureImpl.java:100]: Yes https://github.com/jenkinsci/jenkins-test-harness/issues/672 is still on my list of things to look at, but I believe it is only tangentially related to this PR in that they both happen to involve code run from `WorkUnitContext.synchronizeEnd`; this PR is about the `executors` list, which is used only if `Future.cancel` is called, and does not interact directly with the `Future` done condition.

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
